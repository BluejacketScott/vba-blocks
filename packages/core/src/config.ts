import { join } from 'path';
import { homedir } from 'os';
import { PackageInfo } from './packages';

/**
 * Keep track of absolute paths to various resources and package resolution
 * 
 * Currently hardcoded, but I imagine an .rc or .config.js approach in the future
 */

export interface Config {
  /**
   * cwd for processing
   */
  cwd: string;

  /**
   * absolute path to build directory
   * (build contains binary output of build process)
   */
  build: string;

  /**
   * absolute path to scripts directory
   * (contains applescript and vbs)
   */
  scripts: string;

  /**
   * absolute path to addins directory
   */
  addins: string;

  /**
   * absolute path to cache directory
   * (default is .vba-blocks in home directory)
   */
  cache: string;

  /**
   * absolute path to local registry
   * and url for git remote
   */
  registry: { local: string; remote: string };

  /**
   * resolve package from remote source
   * (default is packages.vba-blocks.com/...)
   */
  resolveRemotePackage: (pkg: PackageInfo) => string;

  /**
   * resolve package locally
   * (default is in {cache}/packages/...)
   */
  resolveLocalPackage: (pkg: PackageInfo) => string;

  /**
   * resolve expanded package ("source")
   * (default is in {cache}/sources/...)
   */
  resolveSource: (pkg: PackageInfo) => string;
}

export async function loadConfig(): Promise<Config> {
  const cwd = process.cwd();
  const build = join(cwd, 'build');
  const scripts = join(__dirname, '../scripts');
  const addins = join(__dirname, '../../addin/build');
  const cache = join(homedir(), '.vba-blocks');
  const registry = {
    local: join(cache, 'registry'),
    remote: 'https://github.com/vba-blocks/registry.git'
  };

  const resolveRemotePackage = pkg =>
    `https://packages.vba-blocks.com/${pkg.name}/v${pkg.version}.tar.gz`;
  const resolveLocalPackage = pkg =>
    join(cache, `packages/${pkg.name}/v${pkg.version}.tar.gz`);
  const resolveSource = pkg =>
    join(cache, `sources/${pkg.name}/v${pkg.version}`);

  const config: Config = {
    cwd,
    build,
    scripts,
    addins,
    cache,
    registry,
    resolveRemotePackage,
    resolveLocalPackage,
    resolveSource
  };

  return config;
}
