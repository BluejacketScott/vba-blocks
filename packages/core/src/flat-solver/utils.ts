export type Iterable = any;

export function forEach(
  iterator: Iterable,
  fn: (item: any, index: number) => any
) {
  let index = 0;
  for (const item of iterator) {
    fn(item, index++);
  }
}

export function map(
  iterator: Iterable,
  fn: (item: any, index: number) => any
): any[] {
  let index = 0;
  const results = [];
  for (const item of iterator) {
    results.push(fn(item, index++));
  }
  return results;
}

export function pluck(values: Iterable, key: string): any[] {
  return map(values, value => value && value[key]);
}

export function reduce<T>(
  iterator: Iterable,
  fn: (memo: T, item: any) => T,
  memo: T
) {
  for (const item of iterator) {
    memo = fn(memo, item);
  }

  return memo;
}

export function unique(values: Iterable): any[] {
  const asSet = new Set(values);
  return Array.from(asSet);
}
