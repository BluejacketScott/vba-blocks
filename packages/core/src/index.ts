export {
  Version,
  Dependency,
  Feature,
  Reference,
  Source,
  Target,
  Manifest
} from './manifest';
export { Config } from './config';

export { default as resolve } from './resolve';
