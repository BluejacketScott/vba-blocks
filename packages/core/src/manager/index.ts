export { Registration, getRegistrationId } from './registration';
export { default as registry } from './registry-manager';
export { default as path } from './path-manager';
export { default as git } from './git-manager';
