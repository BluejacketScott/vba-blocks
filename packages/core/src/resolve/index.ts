import { Config } from '../config';
import { Manifest } from '../manifest';
import { loadLockfile, writeLockfile } from './lockfile';

export interface DependencyGraph {}

export default async function resolve(
  config: Config,
  manifest: Manifest
): Promise<DependencyGraph> {
  return {};
}
