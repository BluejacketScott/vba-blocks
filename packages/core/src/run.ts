import { join } from 'path';
import { Config, loadConfig } from './config';

const extensions = {
  excel: ['xlsx', 'xlsm', 'xlam']
};
const addins = {
  excel: 'vba-blocks.xlam'
};

export default async function run(command, target, args) {
  let application;
  for (let key in extensions) {
    if (extensions[key].includes(target.type)) {
      application = key;
      break;
    }
  }

  if (!application) {
    throw new Error(`Unsupported target type: ${target.type}`);
  }

  const config = await loadConfig();
  const addin = join(config.addins, addins[application]);
  target = join(config.build, `${target.name}.${target.type}`);

  console.log(
    `run(${application}, ${addin}, ${command}, ${JSON.stringify(
      target
    )}, ${JSON.stringify(args)})`
  );
}
