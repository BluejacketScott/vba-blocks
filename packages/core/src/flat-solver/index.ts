import * as semver from 'semver';
import * as Logic from 'logic-solver';
import { forEach, map, pluck } from './utils';
import {
  resolveDependencies,
  FlatGraph,
  FlatGraphNode
} from './resolve-dependencies';

export { resolveDependencies, FlatGraph, FlatGraphNode };

export type Version = string;
export type Dependency = { name: string; version: Version };

export interface Resolver {
  (name: string): Promise<PackageInfo[]>;
}

export interface PackageInfo {
  id?: string;
  name: string;
  version: Version;
  dependencies: Dependency[];
}

export interface Solution {
  success: boolean;
  result?: string[];
}

/**
 * Solve for the resolved flat dependency tree given a set of root dependencies
 */
export async function solve(
  dependencies: Dependency[] | { [name: string]: Version },
  resolver: Resolver
): Promise<Solution> {
  if (!Array.isArray(dependencies))
    dependencies = toArray<Dependency>(dependencies);

  const graph = await resolveDependencies(dependencies, resolver);
  const solver = new Logic.Solver();

  forEach(graph.values(), node => {
    const { name, versions, required } = node;
    const ids = pluck(versions, 'id');

    if (required) {
      solver.require(Logic.exactlyOne(...ids));
    } else {
      solver.require(Logic.atMostOne(...ids));
    }
  });

  const reversed = [...graph.values()].reverse();
  forEach(reversed, node => {
    const { versions } = node;

    forEach(versions, info => {
      const { id } = info;
      forEach(info.dependencies, dependency => {
        const matching = getMatching(dependency, graph);
        solver.require(Logic.implies(id, Logic.or(...matching)));
      });
    });
  });

  const solution = solver.solve();
  if (!solution) return { success: false };

  return {
    success: true,
    result: solution.getTrueVars()
  };
}

function getMatching(dependency, graph) {
  const { name, version: req } = dependency;
  const versions = graph.get(name).versions;

  const matching = versions
    .filter(({ version }) => semver.satisfies(version, req))
    .map(version => version.id);

  return matching;
}

function toArray<T>(dependencies): T[] {
  return map(Object.entries(dependencies), ([name, version]) => {
    return { name, version };
  });
}
