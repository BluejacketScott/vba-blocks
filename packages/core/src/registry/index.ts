import { join, dirname, basename } from 'path';
import * as readline from 'readline';
import { createReadStream, ensureDir, exists } from 'fs-extra';
import { clone, pull } from '../utils/git';
import { Config } from '../config';
import { Version, Dependency } from '../manifest';

export interface Registration {
  name: string;
  version: Version;
  dependencies: Dependency[];
  features: { [name: string]: string[] };
  checksum?: string;
  yanked?: boolean;
}

export async function getRegistered(
  config: Config,
  name: string
): Promise<Registration[]> {
  return new Promise<Registration[]>((resolve, reject) => {
    const path = getPath(config, name);
    const registrations: Registration[] = [];

    const reader = readline.createInterface({
      input: createReadStream(path)
    });
    reader.on('line', line => {
      const { name, vers, deps, cksum, features, yanked } = JSON.parse(line);
      const dependencies = deps.map(dependency => {
        const { name, req, features, optional, default_features } = dependency;
        return { name, range: req, features, optional, default_features };
      });

      registrations.push({
        name,
        version: vers,
        dependencies,
        features,
        checksum: cksum,
        yanked
      });
    });
    reader.on('close', () => resolve(registrations));
    reader.on('error', reject);
  });
}

export async function updateRegistry(config: Config): Promise<void> {
  const { local, remote } = config.registry;

  if (!await exists(local)) {
    const dir = dirname(local);
    await ensureDir(dir);
    await clone(remote, basename(local), dir);
  }

  await pull(local);
}

export function getPath(config: Config, name: string): string {
  let parts;
  if (name.length === 1) {
    parts = [1, name];
  } else if (name.length === 2) {
    parts = [2, name];
  } else if (name.length === 3) {
    parts = [3, name.substring(0, 1)];
  } else {
    parts = [name.substring(0, 2), name.substring(2, 4)];
  }

  return join(config.registry.local, ...parts, name);
}
