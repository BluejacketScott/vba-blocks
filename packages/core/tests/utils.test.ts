import { convertToToml } from '../src/utils';

test('convertToToml', () => {
  expect(
    convertToToml({
      package: [
        {
          name: 'a',
          version: '1.2.3',
          source: 'registry+...',
          checksum: '...',
          dependencies: ['d 0.0.0 (registry+...)']
        },
        {
          name: 'b',
          version: '4.5.6'
        },
        {
          name: 'c',
          version: '7.8.9',
          source: 'git+...',
          checksum: '<none>'
        },
        {
          name: 'd',
          version: '0.0.0',
          source: 'registry+...',
          checksum: '...'
        }
      ]
    })
  ).toMatchSnapshot();
});
