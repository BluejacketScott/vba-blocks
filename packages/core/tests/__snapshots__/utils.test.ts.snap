// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`convertToToml 1`] = `
"[[package]]
name = \\"a\\"
version = \\"1.2.3\\"
source = \\"registry+...\\"
checksum = \\"...\\"
dependencies = [
  \\"d 0.0.0 (registry+...)\\"
]

[[package]]
name = \\"b\\"
version = \\"4.5.6\\"

[[package]]
name = \\"c\\"
version = \\"7.8.9\\"
source = \\"git+...\\"
checksum = \\"<none>\\"

[[package]]
name = \\"d\\"
version = \\"0.0.0\\"
source = \\"registry+...\\"
checksum = \\"...\\"

"
`;
